### PR TITLE
Possibility to restrict the type of features you can select

### PR DESCRIPTION
**Pull request by acanimal**
See #3257 issue.

This PR includes samples and also an initial implementation based on a new `filter` property which accepts a `ol.interaction.SelectFilterFunction` invoked for each selected feature and that determines if the feature must be selected or not.

### CODE REVIEW

File: src/ol/interaction/selectinteraction.js [hunk 1]
```diff
@@ -68,6 +77,12 @@ ol.interaction.Select = function(opt_options) {
    */
   this.multi_ = goog.isDef(options.multi) ? options.multi : false;
 
+  /**
+   * @private
+   * @type {ol.interaction.SelectFilterFunction}
+   */
+  this.filter_ = goog.isDef(options.filter) ? options.filter : null;
```
Comment (elemoine) [inline src/ol/interaction/selectinteraction.js]: The new `filter` option needs to be added to `olx.interaction.SelectOptions`. See the `externs/olx.js` file.

File: src/ol/interaction/selectinteraction.js [hunk 2]
```diff
@@ -13,6 +13,15 @@ goog.require('ol.interaction.Interaction');
 goog.require('ol.style.Style');
 
 
+/**
+ * A function that takes an {@link ol.Feature}. The function should return true
+ * if the feature must be selected or false otherwise.
```
Comment (elemoine) [inline src/ol/interaction/selectinteraction.js]: must -> may

File: src/ol/interaction/selectinteraction.js (hunk 1 above)
Comment (elemoine) [inline src/ol/interaction/selectinteraction.js]: You could change this to:

```js
  this.filter_ = goog.isDef(options.filter) ? options.filter : goog.functions.TRUE;
```

and thereby avoid testing that `this.filter_` is defined in the `handleEvent` function.

File: src/ol/interaction/selectinteraction.js [hunk 3]
```diff
@@ -150,7 +166,10 @@ ol.interaction.Select.handleEvent = function(mapBrowserEvent) {
          * @param {ol.layer.Layer} layer Layer.
          */
         function(feature, layer) {
-          selected.push(feature);
+          if (!goog.isDef(filter) || goog.isFunction(filter) &&
+              filter.call(this, feature)) {
```
Comment (elemoine) [inline src/ol/interaction/selectinteraction.js]: Why do you call the `filter` with the interaction as `this`?

Comment (elemoine) [inline src/ol/interaction/selectinteraction.js]: Note that we do call the `layerFilter` with the interaction as `this` so I don't know why we'd do it for the `filter`.

File: src/ol/interaction/selectinteraction.js (hunk 1 above)
Comment (acanimal) [inline src/ol/interaction/selectinteraction.js]: Nice tip. I just start reading "Closure the definite guide".

File: src/ol/interaction/selectinteraction.js (hunk 3 above)
Comment (acanimal) [inline src/ol/interaction/selectinteraction.js]: There was no reason. I changed it.

File: src/ol/interaction/selectinteraction.js [hunk 4]
```diff
@@ -150,7 +167,9 @@ ol.interaction.Select.handleEvent = function(mapBrowserEvent) {
          * @param {ol.layer.Layer} layer Layer.
          */
         function(feature, layer) {
-          selected.push(feature);
+          if (goog.isFunction(filter) && filter(feature, layer)) {
```
Comment (elemoine) [inline src/ol/interaction/selectinteraction.js]: You don't need the `isFunction` test.

File: src/ol/interaction/selectinteraction.js [hunk 5]
```diff
@@ -140,6 +156,7 @@ ol.interaction.Select.handleEvent = function(mapBrowserEvent) {
   var features = this.featureOverlay_.getFeatures();
   var /** @type {Array.<ol.Feature>} */ deselected = [];
   var /** @type {Array.<ol.Feature>} */ selected = [];
+  var filter = this.filter_;
```
Comment (elemoine) [inline src/ol/interaction/selectinteraction.js]: For better readability I'd remove the use of a local `filter` variable in this case.

**Comment by acanimal** [comment]
Updated PR with the @elemoine comments. Added documentation to the `olx.js` file too.

Although the `./build.py ci` task runs fine I don't know why the apidoc doesn't show the new `filter` property.

In addition, I have added a second parameter to the `ol.interaction.SelectFilterFunction` that references the layers which feature belongs. By some comment on the list I think having a reference to the layer can help in some situations.

File: examples/select-features-with-filter.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+};
+
+// select interaction
+var selectInteraction = new ol.interaction.Select({
+  filter: function(feature, layer) {
+    if (types === null) {
+      return true;
+    } else {
+      return goog.array.contains(types, feature.getGeometry().getType());
```
Comment (elemoine) [inline examples/select-features-with-filter.js]: You can't use `goog` functions in ol3 examples.

**Comment by elemoine** [comment]
@acanimal, I don't think we need a new example for that new option.

File: externs/olx.js [hunk 7]
```diff
@@ -2606,6 +2607,14 @@ olx.interaction.SelectOptions.prototype.toggleCondition;
  */
 olx.interaction.SelectOptions.prototype.multi;
 
+/**
+ * A function that takes an {@link ol.Feature} and an {ol.layer.Layer} and
+ * returns true if the feature may be selected or false otherwise.
+ * @typedef {function(ol.Feature, ol.layer.Layer): boolean}
```
Comment (elemoine) [inline externs/olx.js]: `@typedef` - > `@type`.  And why don't you use `ol.interaction.SelectFilterFunction` here too?

We don't have a typedef for the layer filter so I think I'd drop the typedef entirely.

File: examples/select-features-with-filter.js (hunk 6 above)
Comment (acanimal) [inline examples/select-features-with-filter.js]: Changed to use `indexOf`.

File: externs/olx.js (hunk 7 above)
Comment (acanimal) [inline externs/olx.js]: Sorry, mistake copy&paste.

**Comment by acanimal** [comment]
I think sample can help to show how to use the property. Anyway the property is so trivial that can be removed. Simply confirm me and I will remove. I vote for not removing :)

**Comment by elemoine** [comment]
@acanimal, yes, using the new `filter` option is trivial, and we can't add an example for every option we add. So I would remove the example.

**Comment by acanimal** [comment]
Sample removed.

**Comment by elemoine** [comment]
Did you see my other comments?

File: src/ol/interaction/selectinteraction.js (hunk 5 above)
Comment (acanimal) [inline src/ol/interaction/selectinteraction.js]: I can't remove, filter is executed in another context. I need a reference like with `selected` variable.

File: src/ol/interaction/selectinteraction.js (hunk 4 above)
Comment (acanimal) [inline src/ol/interaction/selectinteraction.js]: Removed

**Comment by acanimal** [comment]
Strange, I didn't receive any notification to my email.
I just make the changes.

**Comment by acanimal** [comment]
@ahocevar @elemoine Hi, what happened with this features? I though it would be merged in the 3.3 release :(
I don't know the the process must follow a PR to be accepted but this feature is important to me at work. Could you let me know some feedback?

Thanks.

**Comment by ahocevar** [comment]
@acanimal Sorry for that. Your pull request cannot be merged without conflicts, this is why it did not get merged yet. Also, it did not have a milestone assigned, so I did not wait for it when I created the release yesterday. But since we have a monthly release schedule, I'm assigning it the 3.4.0 milestone now. So if you are able to resolve the conflicts, we can merge your pull request and it will be in the next release.

**Comment by acanimal** [comment]
I though the conflict was due the addition of #3018, `./build.py ci` tasks runs fine and also travis report "green" result.
I will resolve and update the PR.

Thanks

File: src/ol/interaction/selectinteraction.js (hunk 5 above)
Comment (elemoine) [inline src/ol/interaction/selectinteraction.js]: You could just pass `this` as the third arg to `forEachFeatureAtPixel`.

**Comment by acanimal** [comment]
@ahocevar Merge done

**Comment by ahocevar** [comment]
Have you seen @elemoine's last comment?

**Comment by acanimal** [comment]
Sorry for the late reply. Yes, I just uploaded the modified code using `this.filter_` property instead a local variable.

**Comment by tremby** [comment]
This feature is important for a project I'm currently working on. Very much looking forward to it being available in 3.4.0. Thanks to those involved.

**Comment by ahocevar** [comment]
@elemoine, are your concerns addressed now?

**Comment by elemoine** [comment]
LGTM. I just think we should squash the commits into one before merging. I will do that myself this morning, and create a new PR.

File: src/ol/interaction/selectinteraction.js [hunk 8]
```diff
@@ -27,6 +27,15 @@ ol.SelectEventType = {
 };
 
 
+/**
+ * A function that takes an {@link ol.Feature} and an {ol.layer.Layer} and
```
Comment (doronhorwitz) [inline src/ol/interaction/selectinteraction.js]: The link for ```ol.layer.Layer``` should be
```{@link ol.layer.Layer}``` - missing the ```@link```

Comment (elemoine) [inline src/ol/interaction/selectinteraction.js]: You're right. Thanks for catching this.

**Comment by acanimal** [comment]
Hi, 
I'm "disconnected" from the project due time limitation but, why has been removed this PR from 3.5.0? I code it since 3.3 and never released.
Is there any other new functionallity in 3.5 to achieve the same result?

Cheers.

**Comment by bartvde** [comment]
@acanimal it needs a rebase at least, not sure if there are any other reasons.

**Comment by fredj** [comment]
fixed with #3402 ?

**Comment by elemoine** [comment]
> fixed with #3402 ?

Yes.

**Pull request closed without merging**